### PR TITLE
[Admin][Promotion] Remove leftover rules after their assigned product gets deleted

### LINKED ISSUE #11764: Promotions fails to open if used product is deleted AND new created product with same code as was deleted product appears in all places where was used deleted
**Sylius version affected**: 1.7.6

**Description**  
There can be a situation when there is promotion with rule, which includes resource model in its options (like Contains product, or other rule with Taxon), and the resource is deleted (product is deleted), then admin cannot open promotion edit form, as there are transformation or resource identifier to resource model, and there is no check if resource exist. so you will get the error page.
In other hand, related bug, when product which was used in some places is deleted and after some time is created product with same code as was old. In this situation new created product appears in all places where was used previously deleted. What actually can leads to the mess.

**TC1:**

1. Any product should exist in shop
2. Create new promotion with rule “Contains product“
3. Select product
4. Save promotion
5. Then go to products list and delete product, which is used in promotion rule
6. go back to promotions list
7. try to open saved promotion.

**Actual results:**

- Error page is shown

**Expected results:**

- Edit page should not be broken
- In all other places if user want to delete something what currently is in use is shown additional message. It will be good to have same logic here.

 

Other related issue:

**Preconditions:**

- Product with code X is created
- Created product is used e.g. in promotions or in any other place

**TC2:**

1. Go to the product list
2. Delete the product X
> Product X disappears visually from all places where it was used (e.g. actions → item fixed discount → Product filter)
3. Create new product with same code X

**Actual results:**

- New created product appears in all places where was used previously deleted product with same code. It can lead to the mess, cause user is able to forgot where previously was used old product. Business requirements can be different for the newly created.

**Expected results:**

- New created product is not used instead of previously deleted.

### PR DESCRIPTION
**Pull request by NoResponseMate**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12 |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | fixes #11764 |
| License         | MIT                                                          |

### CODE REVIEW

File: features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    @ui
+    Scenario: Accessing an edit page of a promotion with a rule that contains a removed product
+        Given there is a promotion "Christmas sale" with "Contains product" rule with products "Mug" and "Cup"
+        When I delete the "Cup" product
+        Then I should be notified that "Christmas sale" promotion has been updated
+        And I should be able to modify a "Christmas sale" promotion
+        And the "Christmas sale" promotion should have "Contains product" rule configured
+
+    @ui
+    Scenario: Accessing an edit page of a promotion with a rule that contains a removed product
```
Comment (coldic3) [inline features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature]:
```suggestion
    Scenario: Accessing an edit page of a promotion with a rule that contains a removed product only
```

File: features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature [hunk 2]
```diff
@@ -0,0 +1,26 @@
+@managing_promotions
+Feature: Accessing an edit page of a promotion with a rule that contains a removed taxon
```
Comment (coldic3) [inline features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature]:
```suggestion
Feature: Accessing an edit page of a promotion with a rule that contains a removed product
```

File: features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    In order to change promotion details
+    As an Administrator
+    I want to be able to access an edit page of a promotion with a rule that contains a removed taxon
```
Comment (coldic3) [inline features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature]:
```suggestion
    I want to be able to access an edit page of a promotion with a rule that contains a removed product
```

File: features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    In order to change promotion details
+    As an Administrator
+    I want to be able to access an edit page of a promotion with a rule that contains a removed product
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has "Mug" and "Cup" products
+        And I am logged in as an administrator
+
+    @ui
```
Comment (GSadee) [inline features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature]:
```suggestion
    @ui @no-api
```

File: features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    Scenario: Accessing an edit page of a promotion with a rule that contains the removed product
+        Given there is a promotion "Christmas sale" with "Contains product" rule with products "Mug" and "Cup"
+        When I delete the "Cup" product
+        Then I should be notified that "Christmas sale" promotion has been updated
+        And I should be able to modify a "Christmas sale" promotion
+        And the "Christmas sale" promotion should have "Contains product" rule configured
+
+    @ui
```
Comment (GSadee) [inline features/promotion/managing_promotions/accessing_promotion_with_rule_that_contains_removed_product.feature]:
```suggestion
    @ui @no-api
```

File: src/Sylius/Behat/Context/Ui/Admin/RemovingProductContext.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        private SharedStorageInterface $sharedStorage,
+        private IndexPageInterface $indexPage,
+    ) {
+    }
+
+    /**
+     * @When I delete the :product product
+     * @When I try to delete the :product product
+     */
+    public function iDeleteProduct(ProductInterface $product)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/RemovingProductContext.php]: Missing return type

File: src/Sylius/Behat/Resources/config/services/contexts/ui.xml [hunk 7]
```diff
@@ -322,10 +322,15 @@
             <argument type="service" id="sylius.behat.page.admin.impersonate_user" />
         </service>
 
+        <service id="sylius.behat.context.ui.admin.removing_products" class="Sylius\Behat\Context\Ui\Admin\RemovingProductContext">
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/services/contexts/ui.xml]: I would be for using here FQCN as a service id, as we do for new contexts, for example, for the catalog promotion

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/ProductDeletionListenerSpec.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        $event->getSubject()->willReturn('subject');
+
+        $this->shouldThrow(\InvalidArgumentException::class)->during('removeProductFromPromotionRules', [$event]);
+    }
+
+    function it_does_nothing_when_rule_updaters_do_not_return_any_codes(
+        RequestStack $requestStack,
+        SessionInterface $session,
+        ProductAwareRuleUpdaterInterface $firstUpdater,
+        ProductAwareRuleUpdaterInterface $secondUpdater,
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/ProductDeletionListenerSpec.php]: I would be for keeping the order from `let` method here:
```suggestion
        RequestStack $requestStack,
        ProductAwareRuleUpdaterInterface $firstUpdater,
        ProductAwareRuleUpdaterInterface $secondUpdater,
        SessionInterface $session,
```
the same below

**Review by GSadee — APPROVED** [review]

**Review by TheMilek — APPROVED** [review]

**Comment by TheMilek** [comment]
Thank you, Jan! :tada:

**Pull request merged**